frame: Enable GenesisConfig in no_std

michalkucharczyk opened the pull request:
This change enables `GenesisConfig` compilation in `no_std` environment.

It also enables `Default` implementation for `GenesisConfig` which will be required for no_std in no-native
runtime world. It must be possible to instantiate default GenesisConfig for pallets and runtime.

Step towards: paritytech/polkadot-sdk#25

michalkucharczyk committed:
frame: Default for GenesisConfig in no_std

`Default` for `GenesisConfig` will be required for no_std in no native
runtime world. It must be possible to instantiate default GenesisConfig
for pallets and runtime.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2807906

michalkucharczyk commented:
bot format

michalkucharczyk commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2808154

bkchr commented:
@michalkucharczyk please change the title and add some more description to the pr. This can break downstream code as `GenesisConfig` starts to compile for `no_std` now, which wasn't the case before. So, please mention this :)

ggwpez commented on frame/alliance/src/lib.rs:
```diff
@@ -407,7 +407,6 @@ pub mod pallet {
 		pub phantom: PhantomData<(T, I)>,
 	}
 
-	#[cfg(feature = "std")]
```
Should be possible to replace the impl here with `derive(DefaultNoBound)`.

michalkucharczyk replied:
done in: https://github.com/paritytech/substrate/pull/14108/commits/8a21eb158b0c03f2213fd8dfcafcaa9d0710397b

ggwpez commented on frame/assets/src/lib.rs:
```diff
@@ -375,7 +375,6 @@ pub mod pallet {
 		pub accounts: Vec<(T::AssetId, T::AccountId, T::Balance)>,
 	}
 
-	#[cfg(feature = "std")]
```
dito and more below.

michalkucharczyk replied:
all done in: https://github.com/paritytech/substrate/pull/14108/commits/8a21eb158b0c03f2213fd8dfcafcaa9d0710397b

ggwpez reviewed (commented):
Eventually we could add a `derive(DefaultNoBound)` to the `genesis_config` macro, but that would be a breaking change.

michalkucharczyk commented:
> @michalkucharczyk please change the title and add some more description to the pr. This can break downstream code as `GenesisConfig` starts to compile for `no_std` now, which wasn't the case before. So, please mention this :)

Initially it was only about `Default` (facepalm).
Done.

michalkucharczyk committed:
hash69 in no_std reverted

michalkucharczyk commented:
> Do we expect to put the conditional compilation of the genesis builders into the WASM binary in `construct_runtime`? Or should every pallet also contain a feature gate for WASM genesis?

The current plan is to not add any feature (or conditional compilation) for genesis config in pallets.

Support for `GenesisConfig` building in runtime (the new runtime API will be introduced for that) will be guarded with new feature. We expect linker to do the job and strip unnecessary `GenesisConfig` related staff if the feature is not enabled.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814036

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814037

michalkucharczyk committed:
derive(DefaultNoBound) for GenesisConfig used when possible

michalkucharczyk committed:
treasury: derive(Default)

michalkucharczyk committed:
Merge remote-tracking branch 'origin/master' into mku-frame-default-for-genesis-config

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814597

michalkucharczyk committed:
Cargo.lock update

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814699

michalkucharczyk committed:
genesis_config: compiler error improved

When std feature is not enabled for pallet, the GenesisConfig will be
defined, but serde::{Serialize,Deserialize} traits will not be
implemented.

The compiler error indicates the reason of latter errors.

This is temporary and serde traits will be enabled with together with
`serde` support in frame.

michalkucharczyk commented on frame/support/procedural/src/pallet/expand/genesis_config.rs:
```diff
@@ -126,7 +135,7 @@ pub fn expand_genesis_config(def: &mut Def) -> proc_macro2::TokenStream {
 						"` does not have the std feature enabled, this will cause the `",
 						$pallet_path,
-						"::GenesisConfig` type to be undefined."
+						"::GenesisConfig` type to not implement serde traits."
```
When std feature is not enabled for pallet, the GenesisConfig will be defined, but `serde::{Serialize,Deserialize}` traits will not be implemented.
    
I've decided to change this compiler error to indicate the reason of latter errors occuring during compilation.
    
This is actually temporary and serde traits will be enabled with together with `serde` support in frame.

michalkucharczyk commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into mku-frame-default-for-genesis-config

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/genesisconfig-in-wasm-runtime-a-step-towards-a-non-native-future/2887/1

michalkucharczyk commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into mku-frame-default-for-genesis-config

ggwpez commented on frame/support/procedural/src/pallet/expand/genesis_config.rs:
```diff
@@ -94,15 +94,24 @@ pub fn expand_genesis_config(def: &mut Def) -> proc_macro2::TokenStream {
+				syn::parse_quote!( #[cfg_attr(feature = "std", serde(bound(deserialize = "")))] ),
+			);
+			attrs.push(
+				syn::parse_quote!( #[cfg_attr(feature = "std", serde(crate = #serde_crate))] ),
```
Maybe you can push them all in one `quote` call?

KiChjang replied:
Indeed, this is possible through binding a new `TokenStream` called `serde_attr_arg` or something similar, and use the `quote!` macro for it.

bkchr replied:
this is `parse_quote!` and will not work.

shawntabrizi reviewed (approved):
awesome

michalkucharczyk commented:
bot merge
